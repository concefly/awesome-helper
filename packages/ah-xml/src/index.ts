export * from './XmlHelper';
export * from './XmlNode';
