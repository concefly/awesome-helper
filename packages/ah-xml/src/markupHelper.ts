import { XmlElementNode, XmlNode, XmlTextNode } from './XmlNode';

export function noteToHTML(
  node: XmlNode,
  opt?: {
    tapAttrs?(attrs: [string, string][], current: XmlNode): [string, string][];
  }
): string {
  const selfCloseTagSet = new Set<string>(['br', 'img']);

  const convertXml2Html = (current: XmlNode): string => {
    if (current instanceof XmlElementNode) {
      const tag = current.tagName;

      let attributes = Array.from(current.attributes.entries());

      if (opt?.tapAttrs) {
        attributes = opt.tapAttrs(attributes, current);
      }

      const attrStr = attributes.map(([_n, _v]) => `${_n}="${_.escape(_v)}"`).join(' ');
      const isSelfClose = selfCloseTagSet.has(tag) && current.flatChildren.length === 0;

      if (isSelfClose) {
        return `<${tag} ${attrStr} />`;
      }

      const childrenStr = current.flatChildren.map(convertXml2Html).join('\n');
      return `<${tag} ${attrStr}>${childrenStr}</${tag}>`;
    }

    if (current instanceof XmlTextNode) return current.text;

    return '';
  };

  return convertXml2Html(node);
}
