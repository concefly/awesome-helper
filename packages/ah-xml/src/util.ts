import { xml2js, Element } from 'xml-js';

export interface ElementWithId extends Element {
  id: string;
  parentId: string;
}

export interface IWalkXmlCtx {
  index: number;
  depth: number;
  parent?: Element;
}

export function walkXml(
  current: Element,
  tap: (c: Element, _ctx: IWalkXmlCtx) => void,
  ctx: IWalkXmlCtx = { index: 0, depth: 0 }
) {
  tap(current, ctx);

  current.elements?.forEach((child, index) => {
    walkXml(child, tap, { parent: current, index, depth: ctx.depth + 1 });
  });
}

export function convertXml2List(input: Element | string) {
  // 复制一遍
  const xml =
    typeof input === 'string'
      ? (xml2js(input, { compact: false }) as Element)
      : (JSON.parse(JSON.stringify(input)) as Element);

  const list: ElementWithId[] = [];

  walkXml(xml, (current, { parent, index, depth }) => {
    const id = [depth, index].join('_');

    // walk 是深度遍历，parent 一定已经被打过 id 了
    const parentId = parent ? (parent as ElementWithId)!.id : '';

    Object.assign(current, { id, parentId });
    list.push(current as ElementWithId);
  });

  return list;
}
