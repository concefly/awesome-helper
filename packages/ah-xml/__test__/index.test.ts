import { XmlHelper } from '../src';
import * as fs from 'fs';

const xmlRaw1 = fs.readFileSync('__test__/xml1.xml', { encoding: 'utf-8' });

describe('index', () => {
  it('XmlHelper', () => {
    const xml = new XmlHelper(xmlRaw1);
    expect(xml.list).toMatchSnapshot();
  });
});
