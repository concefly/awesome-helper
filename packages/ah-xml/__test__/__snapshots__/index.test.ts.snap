// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`index XmlHelper 1`] = `
Array [
  XmlNode {
    "id": "0_0",
    "meta": Object {
      "xmlElement": Object {
        "declaration": Object {
          "attributes": Object {
            "encoding": "ISO-8859-1",
            "version": "1.0",
          },
        },
        "elements": Array [
          Object {
            "comment": " Edited with XML Spy v2007 (http://www.altova.com) ",
            "id": "1_0",
            "parentId": "0_0",
            "type": "comment",
          },
          Object {
            "attributes": Object {
              "title": "菜单",
            },
            "elements": Array [
              Object {
                "attributes": Object {
                  "title": "菜名",
                },
                "elements": Array [
                  Object {
                    "elements": Array [
                      Object {
                        "id": "4_0",
                        "parentId": "3_0",
                        "text": "Belgian Waffles",
                        "type": "text",
                      },
                    ],
                    "id": "3_0",
                    "name": "name",
                    "parentId": "2_0",
                    "type": "element",
                  },
                  Object {
                    "elements": Array [
                      Object {
                        "id": "4_0",
                        "parentId": "3_1",
                        "text": "$5.95",
                        "type": "text",
                      },
                    ],
                    "id": "3_1",
                    "name": "price",
                    "parentId": "2_0",
                    "type": "element",
                  },
                  Object {
                    "elements": Array [
                      Object {
                        "id": "4_0",
                        "parentId": "3_2",
                        "text": "two of our famous Belgian Waffles with plenty of real maple syrup",
                        "type": "text",
                      },
                    ],
                    "id": "3_2",
                    "name": "description",
                    "parentId": "2_0",
                    "type": "element",
                  },
                  Object {
                    "elements": Array [
                      Object {
                        "id": "4_0",
                        "parentId": "3_3",
                        "text": "650",
                        "type": "text",
                      },
                    ],
                    "id": "3_3",
                    "name": "calories",
                    "parentId": "2_0",
                    "type": "element",
                  },
                ],
                "id": "2_0",
                "name": "food",
                "parentId": "1_1",
                "type": "element",
              },
              Object {
                "elements": Array [
                  Object {
                    "elements": Array [
                      Object {
                        "id": "4_0",
                        "parentId": "3_0",
                        "text": "Strawberry Belgian Waffles",
                        "type": "text",
                      },
                    ],
                    "id": "3_0",
                    "name": "name",
                    "parentId": "2_1",
                    "type": "element",
                  },
                  Object {
                    "elements": Array [
                      Object {
                        "id": "4_0",
                        "parentId": "3_1",
                        "text": "$7.95",
                        "type": "text",
                      },
                    ],
                    "id": "3_1",
                    "name": "price",
                    "parentId": "2_1",
                    "type": "element",
                  },
                  Object {
                    "elements": Array [
                      Object {
                        "id": "4_0",
                        "parentId": "3_2",
                        "text": "light Belgian waffles covered with strawberries and whipped cream",
                        "type": "text",
                      },
                    ],
                    "id": "3_2",
                    "name": "description",
                    "parentId": "2_1",
                    "type": "element",
                  },
                  Object {
                    "elements": Array [
                      Object {
                        "id": "4_0",
                        "parentId": "3_3",
                        "text": "900",
                        "type": "text",
                      },
                    ],
                    "id": "3_3",
                    "name": "calories",
                    "parentId": "2_1",
                    "type": "element",
                  },
                ],
                "id": "2_1",
                "name": "food",
                "parentId": "1_1",
                "type": "element",
              },
            ],
            "id": "1_1",
            "name": "breakfast_menu",
            "parentId": "0_0",
            "type": "element",
          },
        ],
        "id": "0_0",
        "parentId": "",
      },
    },
    "parentId": "",
    "tree": XmlHelper {
      "getChildrenMap": [Function],
      "getIdMap": [Function],
      "list": [Circular],
      "raw": "<?xml version=\\"1.0\\" encoding=\\"ISO-8859-1\\"?>
<!-- Edited with XML Spy v2007 (http://www.altova.com) -->
<breakfast_menu title=\\"菜单\\">
	<food title=\\"菜名\\">
		<name>Belgian Waffles</name>
		<price>$5.95</price>
		<description>two of our famous Belgian Waffles with plenty of real maple syrup</description>
		<calories>650</calories>
	</food>
	<food>
		<name>Strawberry Belgian Waffles</name>
		<price>$7.95</price>
		<description>light Belgian waffles covered with strawberries and whipped cream</description>
		<calories>900</calories>
	</food>
</breakfast_menu>",
    },
  },
  XmlNode {
    "id": "1_0",
    "meta": Object {
      "xmlElement": Object {
        "comment": " Edited with XML Spy v2007 (http://www.altova.com) ",
        "id": "1_0",
        "parentId": "0_0",
        "type": "comment",
      },
    },
    "parentId": "0_0",
    "tree": XmlHelper {
      "getChildrenMap": [Function],
      "getIdMap": [Function],
      "list": [Circular],
      "raw": "<?xml version=\\"1.0\\" encoding=\\"ISO-8859-1\\"?>
<!-- Edited with XML Spy v2007 (http://www.altova.com) -->
<breakfast_menu title=\\"菜单\\">
	<food title=\\"菜名\\">
		<name>Belgian Waffles</name>
		<price>$5.95</price>
		<description>two of our famous Belgian Waffles with plenty of real maple syrup</description>
		<calories>650</calories>
	</food>
	<food>
		<name>Strawberry Belgian Waffles</name>
		<price>$7.95</price>
		<description>light Belgian waffles covered with strawberries and whipped cream</description>
		<calories>900</calories>
	</food>
</breakfast_menu>",
    },
  },
  XmlNode {
    "id": "1_1",
    "meta": Object {
      "xmlElement": Object {
        "attributes": Object {
          "title": "菜单",
        },
        "elements": Array [
          Object {
            "attributes": Object {
              "title": "菜名",
            },
            "elements": Array [
              Object {
                "elements": Array [
                  Object {
                    "id": "4_0",
                    "parentId": "3_0",
                    "text": "Belgian Waffles",
                    "type": "text",
                  },
                ],
                "id": "3_0",
                "name": "name",
                "parentId": "2_0",
                "type": "element",
              },
              Object {
                "elements": Array [
                  Object {
                    "id": "4_0",
                    "parentId": "3_1",
                    "text": "$5.95",
                    "type": "text",
                  },
                ],
                "id": "3_1",
                "name": "price",
                "parentId": "2_0",
                "type": "element",
              },
              Object {
                "elements": Array [
                  Object {
                    "id": "4_0",
                    "parentId": "3_2",
                    "text": "two of our famous Belgian Waffles with plenty of real maple syrup",
                    "type": "text",
                  },
                ],
                "id": "3_2",
                "name": "description",
                "parentId": "2_0",
                "type": "element",
              },
              Object {
                "elements": Array [
                  Object {
                    "id": "4_0",
                    "parentId": "3_3",
                    "text": "650",
                    "type": "text",
                  },
                ],
                "id": "3_3",
                "name": "calories",
                "parentId": "2_0",
                "type": "element",
              },
            ],
            "id": "2_0",
            "name": "food",
            "parentId": "1_1",
            "type": "element",
          },
          Object {
            "elements": Array [
              Object {
                "elements": Array [
                  Object {
                    "id": "4_0",
                    "parentId": "3_0",
                    "text": "Strawberry Belgian Waffles",
                    "type": "text",
                  },
                ],
                "id": "3_0",
                "name": "name",
                "parentId": "2_1",
                "type": "element",
              },
              Object {
                "elements": Array [
                  Object {
                    "id": "4_0",
                    "parentId": "3_1",
                    "text": "$7.95",
                    "type": "text",
                  },
                ],
                "id": "3_1",
                "name": "price",
                "parentId": "2_1",
                "type": "element",
              },
              Object {
                "elements": Array [
                  Object {
                    "id": "4_0",
                    "parentId": "3_2",
                    "text": "light Belgian waffles covered with strawberries and whipped cream",
                    "type": "text",
                  },
                ],
                "id": "3_2",
                "name": "description",
                "parentId": "2_1",
                "type": "element",
              },
              Object {
                "elements": Array [
                  Object {
                    "id": "4_0",
                    "parentId": "3_3",
                    "text": "900",
                    "type": "text",
                  },
                ],
                "id": "3_3",
                "name": "calories",
                "parentId": "2_1",
                "type": "element",
              },
            ],
            "id": "2_1",
            "name": "food",
            "parentId": "1_1",
            "type": "element",
          },
        ],
        "id": "1_1",
        "name": "breakfast_menu",
        "parentId": "0_0",
        "type": "element",
      },
    },
    "parentId": "0_0",
    "tree": XmlHelper {
      "getChildrenMap": [Function],
      "getIdMap": [Function],
      "list": [Circular],
      "raw": "<?xml version=\\"1.0\\" encoding=\\"ISO-8859-1\\"?>
<!-- Edited with XML Spy v2007 (http://www.altova.com) -->
<breakfast_menu title=\\"菜单\\">
	<food title=\\"菜名\\">
		<name>Belgian Waffles</name>
		<price>$5.95</price>
		<description>two of our famous Belgian Waffles with plenty of real maple syrup</description>
		<calories>650</calories>
	</food>
	<food>
		<name>Strawberry Belgian Waffles</name>
		<price>$7.95</price>
		<description>light Belgian waffles covered with strawberries and whipped cream</description>
		<calories>900</calories>
	</food>
</breakfast_menu>",
    },
  },
  XmlNode {
    "id": "2_0",
    "meta": Object {
      "xmlElement": Object {
        "attributes": Object {
          "title": "菜名",
        },
        "elements": Array [
          Object {
            "elements": Array [
              Object {
                "id": "4_0",
                "parentId": "3_0",
                "text": "Belgian Waffles",
                "type": "text",
              },
            ],
            "id": "3_0",
            "name": "name",
            "parentId": "2_0",
            "type": "element",
          },
          Object {
            "elements": Array [
              Object {
                "id": "4_0",
                "parentId": "3_1",
                "text": "$5.95",
                "type": "text",
              },
            ],
            "id": "3_1",
            "name": "price",
            "parentId": "2_0",
            "type": "element",
          },
          Object {
            "elements": Array [
              Object {
                "id": "4_0",
                "parentId": "3_2",
                "text": "two of our famous Belgian Waffles with plenty of real maple syrup",
                "type": "text",
              },
            ],
            "id": "3_2",
            "name": "description",
            "parentId": "2_0",
            "type": "element",
          },
          Object {
            "elements": Array [
              Object {
                "id": "4_0",
                "parentId": "3_3",
                "text": "650",
                "type": "text",
              },
            ],
            "id": "3_3",
            "name": "calories",
            "parentId": "2_0",
            "type": "element",
          },
        ],
        "id": "2_0",
        "name": "food",
        "parentId": "1_1",
        "type": "element",
      },
    },
    "parentId": "1_1",
    "tree": XmlHelper {
      "getChildrenMap": [Function],
      "getIdMap": [Function],
      "list": [Circular],
      "raw": "<?xml version=\\"1.0\\" encoding=\\"ISO-8859-1\\"?>
<!-- Edited with XML Spy v2007 (http://www.altova.com) -->
<breakfast_menu title=\\"菜单\\">
	<food title=\\"菜名\\">
		<name>Belgian Waffles</name>
		<price>$5.95</price>
		<description>two of our famous Belgian Waffles with plenty of real maple syrup</description>
		<calories>650</calories>
	</food>
	<food>
		<name>Strawberry Belgian Waffles</name>
		<price>$7.95</price>
		<description>light Belgian waffles covered with strawberries and whipped cream</description>
		<calories>900</calories>
	</food>
</breakfast_menu>",
    },
  },
  XmlNode {
    "id": "3_0",
    "meta": Object {
      "xmlElement": Object {
        "elements": Array [
          Object {
            "id": "4_0",
            "parentId": "3_0",
            "text": "Belgian Waffles",
            "type": "text",
          },
        ],
        "id": "3_0",
        "name": "name",
        "parentId": "2_0",
        "type": "element",
      },
    },
    "parentId": "2_0",
    "tree": XmlHelper {
      "getChildrenMap": [Function],
      "getIdMap": [Function],
      "list": [Circular],
      "raw": "<?xml version=\\"1.0\\" encoding=\\"ISO-8859-1\\"?>
<!-- Edited with XML Spy v2007 (http://www.altova.com) -->
<breakfast_menu title=\\"菜单\\">
	<food title=\\"菜名\\">
		<name>Belgian Waffles</name>
		<price>$5.95</price>
		<description>two of our famous Belgian Waffles with plenty of real maple syrup</description>
		<calories>650</calories>
	</food>
	<food>
		<name>Strawberry Belgian Waffles</name>
		<price>$7.95</price>
		<description>light Belgian waffles covered with strawberries and whipped cream</description>
		<calories>900</calories>
	</food>
</breakfast_menu>",
    },
  },
  XmlNode {
    "id": "4_0",
    "meta": Object {
      "xmlElement": Object {
        "id": "4_0",
        "parentId": "3_0",
        "text": "Belgian Waffles",
        "type": "text",
      },
    },
    "parentId": "3_0",
    "tree": XmlHelper {
      "getChildrenMap": [Function],
      "getIdMap": [Function],
      "list": [Circular],
      "raw": "<?xml version=\\"1.0\\" encoding=\\"ISO-8859-1\\"?>
<!-- Edited with XML Spy v2007 (http://www.altova.com) -->
<breakfast_menu title=\\"菜单\\">
	<food title=\\"菜名\\">
		<name>Belgian Waffles</name>
		<price>$5.95</price>
		<description>two of our famous Belgian Waffles with plenty of real maple syrup</description>
		<calories>650</calories>
	</food>
	<food>
		<name>Strawberry Belgian Waffles</name>
		<price>$7.95</price>
		<description>light Belgian waffles covered with strawberries and whipped cream</description>
		<calories>900</calories>
	</food>
</breakfast_menu>",
    },
  },
  XmlNode {
    "id": "3_1",
    "meta": Object {
      "xmlElement": Object {
        "elements": Array [
          Object {
            "id": "4_0",
            "parentId": "3_1",
            "text": "$5.95",
            "type": "text",
          },
        ],
        "id": "3_1",
        "name": "price",
        "parentId": "2_0",
        "type": "element",
      },
    },
    "parentId": "2_0",
    "tree": XmlHelper {
      "getChildrenMap": [Function],
      "getIdMap": [Function],
      "list": [Circular],
      "raw": "<?xml version=\\"1.0\\" encoding=\\"ISO-8859-1\\"?>
<!-- Edited with XML Spy v2007 (http://www.altova.com) -->
<breakfast_menu title=\\"菜单\\">
	<food title=\\"菜名\\">
		<name>Belgian Waffles</name>
		<price>$5.95</price>
		<description>two of our famous Belgian Waffles with plenty of real maple syrup</description>
		<calories>650</calories>
	</food>
	<food>
		<name>Strawberry Belgian Waffles</name>
		<price>$7.95</price>
		<description>light Belgian waffles covered with strawberries and whipped cream</description>
		<calories>900</calories>
	</food>
</breakfast_menu>",
    },
  },
  XmlNode {
    "id": "4_0",
    "meta": Object {
      "xmlElement": Object {
        "id": "4_0",
        "parentId": "3_1",
        "text": "$5.95",
        "type": "text",
      },
    },
    "parentId": "3_1",
    "tree": XmlHelper {
      "getChildrenMap": [Function],
      "getIdMap": [Function],
      "list": [Circular],
      "raw": "<?xml version=\\"1.0\\" encoding=\\"ISO-8859-1\\"?>
<!-- Edited with XML Spy v2007 (http://www.altova.com) -->
<breakfast_menu title=\\"菜单\\">
	<food title=\\"菜名\\">
		<name>Belgian Waffles</name>
		<price>$5.95</price>
		<description>two of our famous Belgian Waffles with plenty of real maple syrup</description>
		<calories>650</calories>
	</food>
	<food>
		<name>Strawberry Belgian Waffles</name>
		<price>$7.95</price>
		<description>light Belgian waffles covered with strawberries and whipped cream</description>
		<calories>900</calories>
	</food>
</breakfast_menu>",
    },
  },
  XmlNode {
    "id": "3_2",
    "meta": Object {
      "xmlElement": Object {
        "elements": Array [
          Object {
            "id": "4_0",
            "parentId": "3_2",
            "text": "two of our famous Belgian Waffles with plenty of real maple syrup",
            "type": "text",
          },
        ],
        "id": "3_2",
        "name": "description",
        "parentId": "2_0",
        "type": "element",
      },
    },
    "parentId": "2_0",
    "tree": XmlHelper {
      "getChildrenMap": [Function],
      "getIdMap": [Function],
      "list": [Circular],
      "raw": "<?xml version=\\"1.0\\" encoding=\\"ISO-8859-1\\"?>
<!-- Edited with XML Spy v2007 (http://www.altova.com) -->
<breakfast_menu title=\\"菜单\\">
	<food title=\\"菜名\\">
		<name>Belgian Waffles</name>
		<price>$5.95</price>
		<description>two of our famous Belgian Waffles with plenty of real maple syrup</description>
		<calories>650</calories>
	</food>
	<food>
		<name>Strawberry Belgian Waffles</name>
		<price>$7.95</price>
		<description>light Belgian waffles covered with strawberries and whipped cream</description>
		<calories>900</calories>
	</food>
</breakfast_menu>",
    },
  },
  XmlNode {
    "id": "4_0",
    "meta": Object {
      "xmlElement": Object {
        "id": "4_0",
        "parentId": "3_2",
        "text": "two of our famous Belgian Waffles with plenty of real maple syrup",
        "type": "text",
      },
    },
    "parentId": "3_2",
    "tree": XmlHelper {
      "getChildrenMap": [Function],
      "getIdMap": [Function],
      "list": [Circular],
      "raw": "<?xml version=\\"1.0\\" encoding=\\"ISO-8859-1\\"?>
<!-- Edited with XML Spy v2007 (http://www.altova.com) -->
<breakfast_menu title=\\"菜单\\">
	<food title=\\"菜名\\">
		<name>Belgian Waffles</name>
		<price>$5.95</price>
		<description>two of our famous Belgian Waffles with plenty of real maple syrup</description>
		<calories>650</calories>
	</food>
	<food>
		<name>Strawberry Belgian Waffles</name>
		<price>$7.95</price>
		<description>light Belgian waffles covered with strawberries and whipped cream</description>
		<calories>900</calories>
	</food>
</breakfast_menu>",
    },
  },
  XmlNode {
    "id": "3_3",
    "meta": Object {
      "xmlElement": Object {
        "elements": Array [
          Object {
            "id": "4_0",
            "parentId": "3_3",
            "text": "650",
            "type": "text",
          },
        ],
        "id": "3_3",
        "name": "calories",
        "parentId": "2_0",
        "type": "element",
      },
    },
    "parentId": "2_0",
    "tree": XmlHelper {
      "getChildrenMap": [Function],
      "getIdMap": [Function],
      "list": [Circular],
      "raw": "<?xml version=\\"1.0\\" encoding=\\"ISO-8859-1\\"?>
<!-- Edited with XML Spy v2007 (http://www.altova.com) -->
<breakfast_menu title=\\"菜单\\">
	<food title=\\"菜名\\">
		<name>Belgian Waffles</name>
		<price>$5.95</price>
		<description>two of our famous Belgian Waffles with plenty of real maple syrup</description>
		<calories>650</calories>
	</food>
	<food>
		<name>Strawberry Belgian Waffles</name>
		<price>$7.95</price>
		<description>light Belgian waffles covered with strawberries and whipped cream</description>
		<calories>900</calories>
	</food>
</breakfast_menu>",
    },
  },
  XmlNode {
    "id": "4_0",
    "meta": Object {
      "xmlElement": Object {
        "id": "4_0",
        "parentId": "3_3",
        "text": "650",
        "type": "text",
      },
    },
    "parentId": "3_3",
    "tree": XmlHelper {
      "getChildrenMap": [Function],
      "getIdMap": [Function],
      "list": [Circular],
      "raw": "<?xml version=\\"1.0\\" encoding=\\"ISO-8859-1\\"?>
<!-- Edited with XML Spy v2007 (http://www.altova.com) -->
<breakfast_menu title=\\"菜单\\">
	<food title=\\"菜名\\">
		<name>Belgian Waffles</name>
		<price>$5.95</price>
		<description>two of our famous Belgian Waffles with plenty of real maple syrup</description>
		<calories>650</calories>
	</food>
	<food>
		<name>Strawberry Belgian Waffles</name>
		<price>$7.95</price>
		<description>light Belgian waffles covered with strawberries and whipped cream</description>
		<calories>900</calories>
	</food>
</breakfast_menu>",
    },
  },
  XmlNode {
    "id": "2_1",
    "meta": Object {
      "xmlElement": Object {
        "elements": Array [
          Object {
            "elements": Array [
              Object {
                "id": "4_0",
                "parentId": "3_0",
                "text": "Strawberry Belgian Waffles",
                "type": "text",
              },
            ],
            "id": "3_0",
            "name": "name",
            "parentId": "2_1",
            "type": "element",
          },
          Object {
            "elements": Array [
              Object {
                "id": "4_0",
                "parentId": "3_1",
                "text": "$7.95",
                "type": "text",
              },
            ],
            "id": "3_1",
            "name": "price",
            "parentId": "2_1",
            "type": "element",
          },
          Object {
            "elements": Array [
              Object {
                "id": "4_0",
                "parentId": "3_2",
                "text": "light Belgian waffles covered with strawberries and whipped cream",
                "type": "text",
              },
            ],
            "id": "3_2",
            "name": "description",
            "parentId": "2_1",
            "type": "element",
          },
          Object {
            "elements": Array [
              Object {
                "id": "4_0",
                "parentId": "3_3",
                "text": "900",
                "type": "text",
              },
            ],
            "id": "3_3",
            "name": "calories",
            "parentId": "2_1",
            "type": "element",
          },
        ],
        "id": "2_1",
        "name": "food",
        "parentId": "1_1",
        "type": "element",
      },
    },
    "parentId": "1_1",
    "tree": XmlHelper {
      "getChildrenMap": [Function],
      "getIdMap": [Function],
      "list": [Circular],
      "raw": "<?xml version=\\"1.0\\" encoding=\\"ISO-8859-1\\"?>
<!-- Edited with XML Spy v2007 (http://www.altova.com) -->
<breakfast_menu title=\\"菜单\\">
	<food title=\\"菜名\\">
		<name>Belgian Waffles</name>
		<price>$5.95</price>
		<description>two of our famous Belgian Waffles with plenty of real maple syrup</description>
		<calories>650</calories>
	</food>
	<food>
		<name>Strawberry Belgian Waffles</name>
		<price>$7.95</price>
		<description>light Belgian waffles covered with strawberries and whipped cream</description>
		<calories>900</calories>
	</food>
</breakfast_menu>",
    },
  },
  XmlNode {
    "id": "3_0",
    "meta": Object {
      "xmlElement": Object {
        "elements": Array [
          Object {
            "id": "4_0",
            "parentId": "3_0",
            "text": "Strawberry Belgian Waffles",
            "type": "text",
          },
        ],
        "id": "3_0",
        "name": "name",
        "parentId": "2_1",
        "type": "element",
      },
    },
    "parentId": "2_1",
    "tree": XmlHelper {
      "getChildrenMap": [Function],
      "getIdMap": [Function],
      "list": [Circular],
      "raw": "<?xml version=\\"1.0\\" encoding=\\"ISO-8859-1\\"?>
<!-- Edited with XML Spy v2007 (http://www.altova.com) -->
<breakfast_menu title=\\"菜单\\">
	<food title=\\"菜名\\">
		<name>Belgian Waffles</name>
		<price>$5.95</price>
		<description>two of our famous Belgian Waffles with plenty of real maple syrup</description>
		<calories>650</calories>
	</food>
	<food>
		<name>Strawberry Belgian Waffles</name>
		<price>$7.95</price>
		<description>light Belgian waffles covered with strawberries and whipped cream</description>
		<calories>900</calories>
	</food>
</breakfast_menu>",
    },
  },
  XmlNode {
    "id": "4_0",
    "meta": Object {
      "xmlElement": Object {
        "id": "4_0",
        "parentId": "3_0",
        "text": "Strawberry Belgian Waffles",
        "type": "text",
      },
    },
    "parentId": "3_0",
    "tree": XmlHelper {
      "getChildrenMap": [Function],
      "getIdMap": [Function],
      "list": [Circular],
      "raw": "<?xml version=\\"1.0\\" encoding=\\"ISO-8859-1\\"?>
<!-- Edited with XML Spy v2007 (http://www.altova.com) -->
<breakfast_menu title=\\"菜单\\">
	<food title=\\"菜名\\">
		<name>Belgian Waffles</name>
		<price>$5.95</price>
		<description>two of our famous Belgian Waffles with plenty of real maple syrup</description>
		<calories>650</calories>
	</food>
	<food>
		<name>Strawberry Belgian Waffles</name>
		<price>$7.95</price>
		<description>light Belgian waffles covered with strawberries and whipped cream</description>
		<calories>900</calories>
	</food>
</breakfast_menu>",
    },
  },
  XmlNode {
    "id": "3_1",
    "meta": Object {
      "xmlElement": Object {
        "elements": Array [
          Object {
            "id": "4_0",
            "parentId": "3_1",
            "text": "$7.95",
            "type": "text",
          },
        ],
        "id": "3_1",
        "name": "price",
        "parentId": "2_1",
        "type": "element",
      },
    },
    "parentId": "2_1",
    "tree": XmlHelper {
      "getChildrenMap": [Function],
      "getIdMap": [Function],
      "list": [Circular],
      "raw": "<?xml version=\\"1.0\\" encoding=\\"ISO-8859-1\\"?>
<!-- Edited with XML Spy v2007 (http://www.altova.com) -->
<breakfast_menu title=\\"菜单\\">
	<food title=\\"菜名\\">
		<name>Belgian Waffles</name>
		<price>$5.95</price>
		<description>two of our famous Belgian Waffles with plenty of real maple syrup</description>
		<calories>650</calories>
	</food>
	<food>
		<name>Strawberry Belgian Waffles</name>
		<price>$7.95</price>
		<description>light Belgian waffles covered with strawberries and whipped cream</description>
		<calories>900</calories>
	</food>
</breakfast_menu>",
    },
  },
  XmlNode {
    "id": "4_0",
    "meta": Object {
      "xmlElement": Object {
        "id": "4_0",
        "parentId": "3_1",
        "text": "$7.95",
        "type": "text",
      },
    },
    "parentId": "3_1",
    "tree": XmlHelper {
      "getChildrenMap": [Function],
      "getIdMap": [Function],
      "list": [Circular],
      "raw": "<?xml version=\\"1.0\\" encoding=\\"ISO-8859-1\\"?>
<!-- Edited with XML Spy v2007 (http://www.altova.com) -->
<breakfast_menu title=\\"菜单\\">
	<food title=\\"菜名\\">
		<name>Belgian Waffles</name>
		<price>$5.95</price>
		<description>two of our famous Belgian Waffles with plenty of real maple syrup</description>
		<calories>650</calories>
	</food>
	<food>
		<name>Strawberry Belgian Waffles</name>
		<price>$7.95</price>
		<description>light Belgian waffles covered with strawberries and whipped cream</description>
		<calories>900</calories>
	</food>
</breakfast_menu>",
    },
  },
  XmlNode {
    "id": "3_2",
    "meta": Object {
      "xmlElement": Object {
        "elements": Array [
          Object {
            "id": "4_0",
            "parentId": "3_2",
            "text": "light Belgian waffles covered with strawberries and whipped cream",
            "type": "text",
          },
        ],
        "id": "3_2",
        "name": "description",
        "parentId": "2_1",
        "type": "element",
      },
    },
    "parentId": "2_1",
    "tree": XmlHelper {
      "getChildrenMap": [Function],
      "getIdMap": [Function],
      "list": [Circular],
      "raw": "<?xml version=\\"1.0\\" encoding=\\"ISO-8859-1\\"?>
<!-- Edited with XML Spy v2007 (http://www.altova.com) -->
<breakfast_menu title=\\"菜单\\">
	<food title=\\"菜名\\">
		<name>Belgian Waffles</name>
		<price>$5.95</price>
		<description>two of our famous Belgian Waffles with plenty of real maple syrup</description>
		<calories>650</calories>
	</food>
	<food>
		<name>Strawberry Belgian Waffles</name>
		<price>$7.95</price>
		<description>light Belgian waffles covered with strawberries and whipped cream</description>
		<calories>900</calories>
	</food>
</breakfast_menu>",
    },
  },
  XmlNode {
    "id": "4_0",
    "meta": Object {
      "xmlElement": Object {
        "id": "4_0",
        "parentId": "3_2",
        "text": "light Belgian waffles covered with strawberries and whipped cream",
        "type": "text",
      },
    },
    "parentId": "3_2",
    "tree": XmlHelper {
      "getChildrenMap": [Function],
      "getIdMap": [Function],
      "list": [Circular],
      "raw": "<?xml version=\\"1.0\\" encoding=\\"ISO-8859-1\\"?>
<!-- Edited with XML Spy v2007 (http://www.altova.com) -->
<breakfast_menu title=\\"菜单\\">
	<food title=\\"菜名\\">
		<name>Belgian Waffles</name>
		<price>$5.95</price>
		<description>two of our famous Belgian Waffles with plenty of real maple syrup</description>
		<calories>650</calories>
	</food>
	<food>
		<name>Strawberry Belgian Waffles</name>
		<price>$7.95</price>
		<description>light Belgian waffles covered with strawberries and whipped cream</description>
		<calories>900</calories>
	</food>
</breakfast_menu>",
    },
  },
  XmlNode {
    "id": "3_3",
    "meta": Object {
      "xmlElement": Object {
        "elements": Array [
          Object {
            "id": "4_0",
            "parentId": "3_3",
            "text": "900",
            "type": "text",
          },
        ],
        "id": "3_3",
        "name": "calories",
        "parentId": "2_1",
        "type": "element",
      },
    },
    "parentId": "2_1",
    "tree": XmlHelper {
      "getChildrenMap": [Function],
      "getIdMap": [Function],
      "list": [Circular],
      "raw": "<?xml version=\\"1.0\\" encoding=\\"ISO-8859-1\\"?>
<!-- Edited with XML Spy v2007 (http://www.altova.com) -->
<breakfast_menu title=\\"菜单\\">
	<food title=\\"菜名\\">
		<name>Belgian Waffles</name>
		<price>$5.95</price>
		<description>two of our famous Belgian Waffles with plenty of real maple syrup</description>
		<calories>650</calories>
	</food>
	<food>
		<name>Strawberry Belgian Waffles</name>
		<price>$7.95</price>
		<description>light Belgian waffles covered with strawberries and whipped cream</description>
		<calories>900</calories>
	</food>
</breakfast_menu>",
    },
  },
  XmlNode {
    "id": "4_0",
    "meta": Object {
      "xmlElement": Object {
        "id": "4_0",
        "parentId": "3_3",
        "text": "900",
        "type": "text",
      },
    },
    "parentId": "3_3",
    "tree": XmlHelper {
      "getChildrenMap": [Function],
      "getIdMap": [Function],
      "list": [Circular],
      "raw": "<?xml version=\\"1.0\\" encoding=\\"ISO-8859-1\\"?>
<!-- Edited with XML Spy v2007 (http://www.altova.com) -->
<breakfast_menu title=\\"菜单\\">
	<food title=\\"菜名\\">
		<name>Belgian Waffles</name>
		<price>$5.95</price>
		<description>two of our famous Belgian Waffles with plenty of real maple syrup</description>
		<calories>650</calories>
	</food>
	<food>
		<name>Strawberry Belgian Waffles</name>
		<price>$7.95</price>
		<description>light Belgian waffles covered with strawberries and whipped cream</description>
		<calories>900</calories>
	</food>
</breakfast_menu>",
    },
  },
]
`;
